content/blog/2016/2016-04-07-pipeline-for-runs-on-hardware.adoc

```diff
@@ -0,0 +1,258 @@
+link:https://wiki.jenkins-ci.org/display/JENKINS/Throttle+Concurrent+Builds+Plugin[Throttle Concurrent Builds], link:https://wiki.jenkins-ci.org/display/JENKINS/Lockable+Resources+Plugin[Lockable Resources]
+ or 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Exclusion-Plugin[Exclusions].
+* Test parallelization on multiple nodes requires using of multiple projects or 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Matrix+Project+Plugin[Matrix configurations], so it causes job chaining again.
+** These build chains can be created via 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Parameterized+Trigger+Plugin[Parameterized Trigger] and 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Copy+Artifact+Plugin[Copy Artifacts], but it complicates job management and build history investigation.
+* Hardware infrastructure is usually flaky. 
+If it fails during the build due to any reason, it's hard to diagnose the issue and re-run the project if the issue comes from hardware.
+** link:https://wiki.jenkins-ci.org/display/JENKINS/Build+Failure+Analyzer[Build Failure Analyzer] allows to identify the root cause of a build failure (e.g. by build log parsing).
+** link:https://wiki.jenkins-ci.org/display/JENKINS/Conditional+BuildStep+Plugin[Conditional Build Step] and 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Flexible+Publish+Plugin[Flexible Publish] plugins allow altering the build flow according to the analysis results.
+** Combination of the plugins above is possible, but it makes job configurations extremely large.
+* Tests on hardware peripherals may take much time. 
+If an infrastructure fails, we may have to restart the run from scratch. 
+So the builds should be robust against infrastructure issues including network failures and Jenkins master restarts.
+* Tests on hardware should be reproducible, so the environment and input parameters should be controlled well.
+** Jenkins supports
+link:https://wiki.jenkins-ci.org/display/JENKINS/Workspace+Cleanup+Plugin[cleaning workspaces], so it can get rid of temporary files generated by previous runs.
+** Jenkins provides support of slaves connected via containers (e.g. 
+  link:https://wiki.jenkins-ci.org/display/JENKINS/Docker+Plugin[Docker]) or VMs, which allow creating clean environments for every new run.
+It's important for 3rd-party tools, which may modify files outside the workspace: user home directory, temporary files, etc.
+** These environments still need to be connected to hardware peripherals, which may be a serious obstacle for Jenkins admins
+
+Usage of 
+link:https://jenkins.io/doc/pipeline/[Jenkins Pipeline]  can effectively address the use-cases above.
+In this blogpost I want to address several cases.
+
+## Pipeline-as-Code for test runs on hardware
+
+Pipeline-as-Code is an emerging technology for representing job configurations as code.
+In Jenkins there are two most popular implementations: Pipeline and JobDSL plugins.
```

Comment:
In Jenkins there are two most popular plugins: 
link:/doc/pipeline/[Pipeline] and link:https://wiki.jenkins-ci.org/display/JENKINS/Job+DSL+Plugin[Job DSL].
JobDSL Plugin internally generates common freestyle jobs according to the script, so it's functionality is similar to the classic approaches.
Pipeline is fundamentally different, because it provides a new engine controlling flows independently from particular nodes and workspaces.
So it provides a higher job description level, which was not available in Jenkins before.